packages/cli/src/scaling/pubsub/pubsub.types.ts

```diff
@@ -1,96 +1,198 @@
+		: { payload: CommandMap[CommandKey] });
+
+	type ToCommand<CommandKey extends keyof CommandMap> = Resolve<_ToCommand<CommandKey>>;
+
+	namespace Command {
+		export type ReloadLicense = ToCommand<'reload-license'>;
+		export type RestartEventBus = ToCommand<'restart-event-bus'>;
+		export type ReloadExternalSecretsProviders = ToCommand<'reload-external-secrets-providers'>;
+		export type CommunityPackageInstall = ToCommand<'community-package-install'>;
+		export type CommunityPackageUpdate = ToCommand<'community-package-update'>;
+		export type CommunityPackageUninstall = ToCommand<'community-package-uninstall'>;
+		export type GetWorkerId = ToCommand<'get-worker-id'>;
+		export type GetWorkerStatus = ToCommand<'get-worker-status'>;
+		export type AddWebhooksTriggersAndPollers = ToCommand<'add-webhooks-triggers-and-pollers'>;
+		export type RemoveTriggersAndPollers = ToCommand<'remove-triggers-and-pollers'>;
+		export type DisplayWorkflowActivation = ToCommand<'display-workflow-activation'>;
+		export type DisplayWorkflowDeactivation = ToCommand<'display-workflow-deactivation'>;
+		export type DisplayWorkflowActivationError = ToCommand<'display-workflow-activation-error'>;
+		export type RelayExecutionLifecycleEvent = ToCommand<'relay-execution-lifecycle-event'>;
+		export type ClearTestWebhooks = ToCommand<'clear-test-webhooks'>;
+	}
+
+	/** Command sent via the `n8n.commands` pubsub channel. */
+	export type Command =
+		| Command.ReloadLicense
+		| Command.RestartEventBus
+		| Command.ReloadExternalSecretsProviders
+		| Command.CommunityPackageInstall
+		| Command.CommunityPackageUpdate
+		| Command.CommunityPackageUninstall
+		| Command.GetWorkerId
+		| Command.GetWorkerStatus
+		| Command.AddWebhooksTriggersAndPollers
+		| Command.RemoveTriggersAndPollers
+		| Command.DisplayWorkflowActivation
+		| Command.DisplayWorkflowDeactivation
+		| Command.DisplayWorkflowActivationError
+		| Command.RelayExecutionLifecycleEvent
+		| Command.ClearTestWebhooks;
+
+	// ----------------------------------
+	//         worker responses
+	// ----------------------------------
+
+	export type WorkerResponseMap = {
+		// #region Lifecycle
+
+		'restart-event-bus': {
+			result: 'success' | 'error';
+			error?: string;
+		};
+
+		'reload-external-secrets-providers': {
+			result: 'success' | 'error';
+			error?: string;
+		};
+
+		// #endregion
+
+		// #region Worker view
+
+		'get-worker-id': never;
+
+		'get-worker-status': WorkerStatus;
+
+		// #endregion
 	};
 
-	// #endregion
-};
+	type _ToWorkerResponse<WorkerResponseKey extends keyof WorkerResponseMap> = {
+		workerId: string;
+		targets?: string[];
+		command: WorkerResponseKey;
+	} & (WorkerResponseMap[WorkerResponseKey] extends never
+		? { payload?: never } // some responses carry no payload
+		: { payload: WorkerResponseMap[WorkerResponseKey] });
+
+	type ToWorkerResponse<WorkerResponseKey extends keyof WorkerResponseMap> = Resolve<
+		_ToWorkerResponse<WorkerResponseKey>
+	>;
```

Comment:
This is a bit hard to read. Especially the `Resolve` bit. Maybe we can simplify this in the future?

Reply:
`Resolve` ensures that the resolved type will be displayed more clearly:

```ts
// like this

type MyCommand = {
	senderId: string;
	targets?: string;
	command: 'add-webhooks-triggers-and-pollers';
	payload: { workflowId: string };
};

// instead of like this

type MyCommand = {
	senderId: string;
	targets?: string[];
	command: 'add-webhooks-triggers-and-pollers';
} & { payload: { workflowId: string } };
```

I chose to have an underscored variant so that I'm not directly applying `Resolve` to the intersection type, but that means having to specify the generics when calling, which makes it a choice between verbosity and code golf.

So `Resolve` and its consumer cannot be simplified further I think.